File: py/offer_10_num_ways.py
class Solution:
    def numWays(self, n: int) -> int:
        # dp n = dp n - 1 + dp n - 2 
        # n = 0 1
        # n = 1 1
        # n = 2 2
        a, b = 1, 1
        i = 2
        while i <= n:
            a, b = b, a+b
            i += 1
        
        return b%1000000007

if __name__ == '__main__':
    s = Solution()
    print(s.numWays(2))

File: offer10_num_ways.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    // n = 0: 1
    // n = 1: 1
    // n >= 2: f(n) = f(n-1) + f(n-2)
    int numWays(int n)
    {
        int i = 1;
        int j = 1;
        if (n == 0)
            return i;
        
        int temp = 0;
        int count = 2;
        while (count <= n)
        {
            count += 1;
            temp = i;
            i = j;
            j = (j + temp) % 1000000007;
        }
        return j;
    }
};

int main()
{
    return 0;
}